### DISCUSSION TITLE
Help: Custom snippets not working

### DISCUSSION THREAD

**Discussion by rubiin** [LazyVim/LazyVim#3216] [General]
The snippets was working fine with luasnip. After recent upgrade, I deleted the luasnip setup I was using in favor of native snippets. And for some reason , the snippet is not working .

**Comment by radvil** [comment]
drop your nvim-cmp config here. Maybe it's not integrated correctly (just my guess).

**Reply by rubiin** [reply]
```lua
local compare = require("cmp.config.compare")
local user_icons = require("custom.icons")

return {
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "petertriho/cmp-git",
      "hrsh7th/cmp-cmdline",
      { "garymjr/nvim-snippets", opts = { friendly_snippets = true, search_paths = { vim.g.vscode_snippets_path } } },
    },
    opts = function(_, opts)
      local cmp = require("cmp")

      opts.confirm_opts = {
        behavior = cmp.ConfirmBehavior.Replace,
        select = false,
      }

      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            -- You could replace select_next_item() with confirm({ select = true }) to get VS Code autocompletion behavior
            cmp.select_next_item()
          elseif vim.snippet.active({ direction = 1 }) then
            vim.schedule(function()
              vim.snippet.jump(1)
            end)
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif vim.snippet.active({ direction = -1 }) then
            vim.schedule(function()
              vim.snippet.jump(-1)
            end)
          else
            fallback()
          end
        end, { "i", "s" }),
      })

      opts.duplicates = {
        nvim_lsp = 1,
        luasnip = 1,
        look = 1,
        cmp_tabnine = 1,
        copilot = 1,
        buffer = 1,
        path = 1,
      }

      opts.completion = {
        completeopt = vim.o.completeopt,
      }
      opts.sorting = {
        comparators = {
          -- Original order: https://github.com/hrsh7th/nvim-cmp/blob/538e37ba87284942c1d76ed38dd497e54e65b891/lua/cmp/config/default.lua#L65-L74
          -- Definitions of compare function https://github.com/hrsh7th/nvim-cmp/blob/main/lua/cmp/config/compare.lua
          compare.offset,
          compare.recently_used, -- higher
          compare.score,
          compare.exact, -- lower
          compare.kind, -- higher (prioritize snippets)
          compare.locality,
          compare.length,
          compare.order,
        },
      }
      opts.window = {
        completion = { border = vim.g.border_style, scrolloff = vim.o.scrolloff, scrollbar = "║" },
        documentation = { border = vim.g.border_style, scrolloff = vim.o.scrolloff, scrollbar = "║" },
        preview = { border = vim.g.border_style, scrolloff = vim.o.scrolloff, scrollbar = "║" },
      }
      opts.formatting = {
        fields = { "kind", "abbr", "menu" },
        format = function(entry, item)
          -- load lspkind icons
          item.kind = string.format(" %s  %s", user_icons.kinds[item.kind], item.kind)

          item.menu = ({
            cmp_tabnine = "[Tabnine]",
            copilot = "[Copilot]",
            spell = "[Spell]",
            buffer = "[Buffer]",
            orgmode = "[Org]",
            look = "[Dictionary]",
            nvim_lsp = "[LSP]",
            git = "[Git]",
            nvim_lua = "[Lua]",
            path = "[Path]",
            tmux = "[Tmux]",
            latex_symbols = "[Latex]",
            luasnip = "[Snippet]",
          })[entry.source.name]

          local label = item.abbr
          local truncated_label = vim.fn.strcharpart(label, 0, 50)
          if truncated_label ~= label then
            item.abbr = truncated_label .. "..."
          end

          return item
        end,
      }

      -- Set configuration for specific filetype.
      cmp.setup.filetype("gitcommit", {
        sources = cmp.config.sources({
          { name = "git" }, -- You can specify the `git` source if [you were installed it](https://github.com/petertriho/cmp-git).
        }, {
          { name = "buffer" },
        }),
      })

      -- LUA: disable annoying `--#region` suggestions
      cmp.setup.filetype("lua", {
        enabled = function()
          local line = vim.api.nvim_get_current_line()
          return not (line:find("%s%-%-?$") or line:find("^%-%-?$"))
        end,
      })

      -- SHELL: disable `\[` suggestions at EoL
      cmp.setup.filetype("sh", {
        enabled = function()
          local col = vim.fn.col(".") - 1
          local charBefore = vim.api.nvim_get_current_line():sub(col, col)
          return charBefore ~= "\\"
        end,
      })
```

The path is correct as well

**Reply by rubiin** [reply]
![image](https://github.com/LazyVim/LazyVim/assets/8222059/10bc8934-2c6b-4bb0-b31f-eee230f67a60)

**Reply by rubiin** [reply]
might be an upstream issue

**Comment by folke** [comment]
Remove all your customizations to cmp and snippets and use the default LazyVim configs.
If that works, you can try adding in your customizations

**Reply by rubiin** [reply]
doing that as we speak

**Reply by rubiin** [reply]
still a issue for custom snippets with a fresh lazy install

**Reply by rubiin** [reply]
figured out the issue :

This does not handle nested folders
https://github.com/garymjr/nvim-snippets#todo
 Add support for friendly-snippets package.json definitions


Had to remove the sub folder
![image](https://github.com/LazyVim/LazyVim/assets/8222059/b16d89f8-9d5a-48bd-9f06-03376c7a045b)

**Reply by dpetka2001** [reply]
@rubiin Are you sure about that? I just tried nested folders and it works for me.

**Reply by rubiin** [reply]
yeah I tried on my setup

**Reply by dpetka2001** [reply]
I have `nvim-snippets` configured like [this](https://github.com/dpetka2001/dotfiles/blob/83e725965227d4c313c8c54272fbcc5bd480c5ba/dot_config/nvim/lua/plugins/coding.lua#L67-L77) and the structure for snippets is [here](https://github.com/dpetka2001/dotfiles/tree/main/dot_config/nvim/misc/snippets) (all the snippets were copied from yours and I just added some simple text printing stuff and just changed the structure a bit to see if it works with nested folders). Look at `package.json` and how I have 2 folders, one for `all` and one for `lua` and `lua` folder has 2 files in it. I can see the snippets of both files in a Lua file as well as the snippets from `all`.

**Reply by rubiin** [reply]
will try to setup a minimal repro when I am free

**Reply by dpetka2001** [reply]
I created a test repo with default LazyVim settings and then added native snippets support. You can see the changes from the default LazyVim installation in this [commit](https://github.com/dpetka2001/test/commit/fbce434eb382222e02f0a23e0e6bede23348c535). Nested folders work for me for custom snippets, as well as `friendly-snippets` tested in Python and Lua files.

**Reply by rubiin** [reply]
lemme check

**Comment by radvil** [comment]
You might wanna try to add this to your config option.

```lua
     table.insert(opts.sources, { name = "snippets" })
      opts.snippet = {
        expand = function(args)
          vim.snippet.expand(args.body)
        end,
      }
```

**Reply by radvil** [reply]
This is based on LazyVim config option itself if you want to use use native snippet, You may try if this can solve it.

**Reply by rubiin** [reply]
the opts are merged with parent I guess thats not necessary

**Reply by dpetka2001** [reply]
These are default options both in [nvim-cmp](https://github.com/hrsh7th/nvim-cmp/blob/5260e5e8ecadaf13e6b82cf867a909f54e15fd07/lua/cmp/config/default.lua#L32-L35) and [nvim-snippets](https://github.com/garymjr/nvim-snippets/blob/f394d17b9a83820714957a06c6ed8e12223f3034/lua/snippets/config/init.lua#L14-L17), so they shouldn't be needed if you use both of them with default settings.

**Comment by Demianeen** [comment]
I had all my luasnip snippets in the `snippets` folder which is the default for nvim-snippets. Because of this I got an error on each opening of neovim.

![telegram-cloud-photo-size-2-5368358633945031309-y](https://github.com/LazyVim/LazyVim/assets/51330172/1f48f031-2295-4e0c-be42-f62652de3ea9)

It was unintuitive for me and I spent half the day fixing it. Hopefully, it will save time for somebody

**Comment by phuongthuan** [comment]
Same with my issue. I'm using LuaSnip (load custom snippets) + nvim-cmp with Super tab configuration. It worked very smooth until upgrade to nvim 0.10.
Here is my configuration.
Does anyone has been resolved this please help me 😞 

It would be perfect if we can migrate LuaSnip to `nvim-snippets` 🥇 


```
  {
    "L3MON4D3/LuaSnip",
    dependencies = {
      {
        "rafamadriz/friendly-snippets",
        config = function()
          require("luasnip.loaders.from_vscode").lazy_load({ paths = vim.fn.stdpath("config") .. "/snippets" })
        end,
      },
    },
    keys = function()
      return {}
    end,
  },
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")
      local cmp = require("cmp")

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_next_item()
            -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
            -- this way you will only jump inside the snippet region
          elseif luasnip.expand_or_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { "i", "s" }),
      })
    end,
  },
```

**Reply by dpetka2001** [reply]
[nvim-snippets](https://github.com/garymjr/nvim-snippets?tab=readme-ov-file#nvim-snippets) only supports VsCode style snippets, not Luasnip style snippets. You would have to migrate your snippets to VsCode snippets syntax. I have created some simple VsCode style snippets in my dotfiles to play around with them and see if I will keep using them or not. You can take a look [here](https://github.com/dpetka2001/dotfiles/tree/main/dot_config/nvim/misc/snippets) for how the structure is.

But if you don't intend to migrate your Luasnip snippets to Vscode style snippets, I'd suggest to keep using Luasnip. It's up to you.

**Reply by folke** [reply]
indeed, just use the luasnip extra

**Reply by Demianeen** [reply]
I see that you use ./snippets/* location. It is also used by default in nvim-snippets. Maybe the issue is with it

**Comment by rubiin** [comment]
https://github.com/garymjr/nvim-snippets/pull/31 working fine after the merge . Just tested
